### ISSUE TITLE
Flaky IT: (Compile=openjdk8, Run=openjdk8) query retry integration test for missing segments

### ISSUE THREAD

**Issue by paul-rogers** [apache/druid#12700]
The following IT failed: [(Compile=openjdk8, Run=openjdk8) query retry integration test for missing segments](https://app.travis-ci.com/github/apache/druid/jobs/574410506). Usual caveats about the PR in question not changing code in the area of the IT.

```text
2022-06-22T22:09:23,719 INFO [main] org.apache.druid.testing.utils.LoggerListener - Starting [org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testWithRetriesDisabledPartialResultAllowed]
2022-06-22T22:09:37,785 ERROR [main] org.apache.druid.testing.utils.LoggerListener - Failed [org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testWithRetriesDisabledPartialResultAllowed]
java.lang.AssertionError: At least one query is expected to return less results. expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:99) ~[testng-7.3.0.jar:?]
	at org.testng.Assert.failNotEquals(Assert.java:1037) ~[testng-7.3.0.jar:?]
	at org.testng.Assert.assertTrue(Assert.java:45) ~[testng-7.3.0.jar:?]
	at org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testQueries(ITQueryRetryTestOnMissingSegments.java:217) ~[test-classes/:?]
	at org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testQueries(ITQueryRetryTestOnMissingSegments.java:141) ~[test-classes/:?]
	at org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testWithRetriesDisabledPartialResultAllowed(ITQueryRetryTestOnMissingSegments.java:122) ~[test-classes/:?]
...

[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 147.022 s <<< FAILURE! - in TestSuite
[ERROR] testWithRetriesDisabledPartialResultAllowed(org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments)  Time elapsed: 14.068 s  <<< FAILURE!
java.lang.AssertionError: At least one query is expected to return less results. expected [true] but found [false]
	at org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testQueries(ITQueryRetryTestOnMissingSegments.java:217)
	at org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testQueries(ITQueryRetryTestOnMissingSegments.java:141)
	at org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments.testWithRetriesDisabledPartialResultAllowed(ITQueryRetryTestOnMissingSegments.java:122)
[ERROR] testWithRetriesDisabledPartialResultDisallowed(org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments)  Time elapsed: 11.52 s
[ERROR] testWithRetriesEnabledPartialResultDisallowed(org.apache.druid.tests.query.ITQueryRetryTestOnMissingSegments)  Time elapsed: 10.446 s
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ITQueryRetryTestOnMissingSegments.testWithRetriesDisabledPartialResultAllowed:122->testQueries:141->testQueries:217 At least one query is expected to return less results. expected [true] but found [false]
```

**Comment by paul-rogers** [comment]
Failed again in [this build](https://app.travis-ci.com/github/apache/druid/jobs/574688587) with the same output as above.

**Comment by rohangarg** [comment]
I looked at the failing IT and have some observations (including test explanation) : 
1. The test group verifies `RetryQueryRunner` feature for missing segments. The feature is needed incase a segment is dropped by a historical after getting a query scheduled by the broker on it but before processing of the segment. 
2. Within that feature, the flaky test verifies the case where retries on missing segments are disallowed but the client is ok with incomplete results.
3. For verification, the test launches a normal historical and a bad historical. The bad historical announces that it can serve all the segments but reports the testing segment as missing whenever it is queried. A testing query on a single segment datasource is run 50 times assuming that the it will run atleast once on both good and bad historical. Finally, the test checks that the out of the 50 runs, there should be atleast one run where the query gave incomplete results and there should be atleast one run where the query gave complete results. Further, there shouldn't be any query failures since the client is OK with incomplete results.
4. The test failure tagged above says that it didn't find a run where the results were incomplete. I tried to reproduce it locally but couldn't do so even after more than 50 automated runs. Possible reason for flakiness could be related to the setup of historicals during tests which interacts with cloud storage as well to get pre-populated data. The test setup only verifies that the testing datasource is fully available. It might be a case that the good historical gets setup fine and the bad historical faces an independent problem, which leads to the testing datasource being fully available but failure in test since there'll be no incomplete results.

For a fix, I think instead of accommodating more things in the exitising test, we can refactor the test as follows : 
1. Remove the good historical setup from the query-retry tests
2. The bad historical reports missing segment for the same query for a configurable number of times. So, if the bad historical reports the segment missing for N times, a retry of N+1 should always make any query succeed with complete results.
3. Refactor the tests for tighter assertions based on the above two changes

Incase the existing setup of two historicals is supposed to verify something in query-retry feature I'm missing, we can explore more solutions keeping the current setup intact.
